x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/SingleUseKey.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.repositories.encrypted;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.common.CheckedSupplier;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.collect.Tuple;
+
+import javax.crypto.SecretKey;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Container class for a {@code SecretKey} with a unique identifier, and a 4-byte wide {@code Integer} nonce, that can be used for a
+ * single encryption operation. Use {@link #createSingleUseKeySupplier(CheckedSupplier)} to obtain a {@code Supplier} that returns
+ * a new {@link SingleUseKey} instance on every invocation. The number of unique {@code SecretKey}s (and their associated identifiers)
+ * generated is minimized and, at the same time, ensuring that a given {@code nonce} is not reused with the same key.
+ */
+final class SingleUseKey {
+    private static final Logger logger = LogManager.getLogger(SingleUseKey.class);
+    static final int MIN_NONCE = Integer.MIN_VALUE;
+    static final int MAX_NONCE = Integer.MAX_VALUE;
+    private static final int MAX_ATTEMPTS = 9;
+    private static final SingleUseKey EXPIRED_KEY = new SingleUseKey(null, null, MAX_NONCE);
+
+    private final BytesReference keyId;
+    private final SecretKey key;
+    private final int nonce;
+
+    // for tests use only!
+    SingleUseKey(BytesReference KeyId, SecretKey Key, int nonce) {
+        this.keyId = KeyId;
+        this.key = Key;
+        this.nonce = nonce;
+    }
+
+    public BytesReference getKeyId() {
+        return keyId;
+    }
+
+    public SecretKey getKey() {
+        return key;
+    }
+
+    public int getNonce() {
+        return nonce;
+    }
+
+    /**
+     * Returns a {@code Supplier} of {@code SingleUseKey}s so that no two instances contain the same key and nonce pair.
+     * A new key is generated only when the {@code nonce} space has been exhausted.
+     */
+    static <T extends Exception> CheckedSupplier<SingleUseKey, T> createSingleUseKeySupplier(
+        CheckedSupplier<Tuple<BytesReference, SecretKey>, T> keyGenerator
+    ) {
+        final AtomicReference<SingleUseKey> keyCurrentlyInUse = new AtomicReference<>(EXPIRED_KEY);
+        return createSingleUseKeySupplier(keyGenerator, keyCurrentlyInUse);
+    }
+
+    // for tests use only
```

Comment:
If this is for tests only, why is it referenced in the class javadoc? I'm a little confused about how this class ought to be used.

Reply:
This method is for test only.
What is referred to in the javadoc is the method with only one parameter.
To alleviate the confusion I have renamed the method that ought to only be used in tests from `createSingleUseKeySupplier` to `internalSingleUseKeySupplier`.